src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/CachedTagHelperResolver.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.AspNetCore.Razor.Telemetry;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor;
+
+internal partial class CachedTagHelperResolver(ITelemetryReporter telemetryReporter)
+{
+    private readonly CompilationTagHelperResolver _innerResolver = new(telemetryReporter);
+    private readonly object _gate = new();
+    private int _currentResultId;
+
+    private record Entry(int ResultId, ImmutableArray<Checksum> Checksums);
+    private readonly MemoryCache<ProjectId, Entry> _projectResultCache = new();
+
+    public ImmutableArray<TagHelperDescriptor> GetValues(ProjectId projectKey, int resultId)
+    {
+        if (!TryGetCachedChecksums(projectKey, resultId, out var cachedChecksums))
+        {
+            return [];
+        }
+
+        using var builder = new PooledArrayBuilder<TagHelperDescriptor>();
+        foreach (var checksum in cachedChecksums)
+        {
+            var value = TryGet(checksum);
+            if (value is null)
+            {
+                continue;
+            }
```

Comment:
Is this ever a valid condition? Is it possible for a `TagHelperDescriptor` to be garbage collected such that the most recent set for a `ProjectId` can no longer be retrieved successfully and a partial set is returned?

Reply:
Unless I'm missing something, all of the data structures hold onto checksums. Is something holding onto the `TagHelperDescriptors` somewhere? If not, isn't it possible for `TagHelperDescriptors` to be garbage collected? Instead of holding onto an `ImmutableArray<Checksum>` on the client side per `ProjectId`, should it hold onto an `ImmutableArray<TagHelperDescriptor>` to ensure that tag helpers are rooted in the client?

In VS, this works because the `ProjectSnapshotManager` is ultimately holding onto active tag helpers through `ProjectWorkspaceState`. However, at this layer, I don't think anything is necessarily doing that. I could totally be missing something though.

Reply:
As this has developed I don't think we even need a cache here. The tag helpers will have to be calculated every time anyways, so we will always be checking against the checksum for the whole project and not just the checksums. This was split initially because I was thinking about sending partial information in a later PR, but even then the cache would have to be updated every time for validation. 

With that, I think this resolver can be removed and a normal resolver can be used.

Reply:
This should be fixed